Fix endless authorization loop (workaround)

drewgillies opened the pull request:
Addresses: https://github.com/NuGet/Engineering/issues/3465

Revving OWIN package versions fixes more stringent requirements by browsers on samesite cookies (described here https://docs.microsoft.com/en-us/aspnet/samesite/owin-samesite). This issue was masking another--roles in our AAD RBAC. Further investigation required there, and this will unbreak us in the meantime.

skofman1 commented on src/NuGet.Status/NuGet.Status.csproj:
```diff
@@ -219,6 +214,24 @@
     <ThirdPartyBinaries Include="Newtonsoft.Json.dll" Visible="false" />
     <ThirdPartyBinaries Include="Owin.dll" Visible="false" />
   </ItemGroup>
+  <ItemGroup>
```
>ItemGroup [](start = 3, length = 9)

please revert changes that are unrelated to the fix

skofman1 commented on src/NuGet.Status/Controllers/AuthorizedController.cs:
```diff
@@ -0,0 +1,138 @@
+
+namespace NuGet.Status.Controllers
+{
+    public class AuthorizedController : AppController
```
>AuthorizedController [](start = 17, length = 20)

The name of the controller doesn't match what it does..

Tratcher replied:
You split these just because of the auth? You really shouldn't need two controllers to make auth work, re-merge them.

Tratcher commented on src/NuGet.Status/App_Start/Startup.cs:
```diff
@@ -57,13 +57,20 @@ public void Configuration(IAppBuilder app)
                         {
+                            RedirectToIdentityProvider = notification =>
+                            {
+                                notification.Options.RedirectUri = new Uri(new Uri(_rootUri), "Authorized").ToString();
```
Remove this, you've overriding global state per-request, and the wrong field at that. If you recombine the controllers then you shouldn't need to change any redirects, it will return back to the original endpoint after authenticating.

drewgillies commented:
Thanks for the input, @Tratcher. I've made those changes here (as you can see), with the following adjustments and reasons:
- _rootUri + "/signin-oidc" results in "https://localhost/signin-oidc", which creates a path which isn't authorized for the app, giving a path error. You may have suggested this because you thought we want auth on that path--but it's actually on "https://localhost/admin" that we want auth (the rest is public). So, I've made it _redirectUri + "/signin-oidc".
- I've remerged the controllers as you've said--I've also removed the HttpGetAttribute from the Index method as it now needs to service both the get and the post (after auth). It doesn't get that far though.
- I've removed the URI change in the redirectToIdentityProvider notification action.

We now have the looping problem again. I wanted you to see this in context of the code though in case something stands out. Thanks again.

drewgillies commented:
@Tratcher update: In case this is what you meant, I've added "https://localhost/signin-oidc" to the accepted paths in the app registration, and added a routing path in the app routing the new URI to the AdminController.Index method, but it's the same issue--it simply won't access the controller without looping until failure.

Tratcher commented:
> and added a routing path in the app routing the new URI to the AdminController.Index method

I don't see that change in your PR. And no, this path should not point at the admin controller, it's an independent resource used for processing one stage of the OIDC handshake. This path is considered anonymous because the authentication process is still incomplete.

RedirectUri should be  "https://localhost/signin-oidc", not "https://localhost/admin/signin-oidc".

drewgillies commented:
@Tratcher ah yes, I didn't push that change. Apologies. Anyway, from what you say it wasn't the correct approach so I've reverted it. I've also now changed the RedirectUri to "https://localhost/signin-oidc" (I still needed to have an accepted path in place for this URI in the app registration as it would bounce with an error otherwise). It still loops though, so I'm clearly missing something.

drewgillies commented:
Discussed offline with @Tratcher (thanks!) and we've determined that the OWIN cookie issue is mitigated by the nupkg version change, but we have a blocker on roles in our RBAC. This will take some further digging, so for now a workaround will be to allowlist admin users via the config (;-separated). When we have roles fixed in AAD, we can return the controller to `[Authorize(Roles="Admins")]`, but for now the controller will be `[Authorize]`. I'll also PR some lists into our configs and update our allowed paths in the web registration to include `signin-oidc`.

joelverhagen commented on src/NuGet.Status/App_Start/Startup.cs:
```diff
@@ -55,15 +54,21 @@ public void Configuration(IAppBuilder app)
+                            RedirectToIdentityProvider = notification =>
+                            {
+                                notification.ProtocolMessage.Prompt = "select_account"; // force entering of credential
+                                return Task.FromResult(0);
```
`Task.CompletedTask`?

drewgillies replied:
Just following convention from below callback, but will change it.

drewgillies replied:
I was wondering about this. It may be worth leaving as-is--every example (as well as the below callback) seems to use `FromResult(0)`: e.g. https://stackoverflow.com/questions/31328378/changing-returnurl-in-owin-redirecttoidentityprovider-notification
...even though the return type is `Task` and not `Task<int>`. Potentially there is some secret sauce afoot here? At any rate, this seems to be the convention. We even do it in gallery code: https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery.Services/Authentication/Providers/AzureActiveDirectoryV2/AzureActiveDirectoryV2Authenticator.cs#L266
...but I can change it if you want. Just doesn't seem the done thing for some reason.

Tratcher replied:
No, it's just that these libraries pre-date Task.CompletedTask and the samples were never updated.

drewgillies replied:
Thanks, @Tratcher!

joelverhagen commented on src/NuGet.Status/Configuration/StatusConfiguration.cs:
```diff
@@ -13,6 +13,8 @@ public class StatusConfiguration : Services.Configuration.Configuration
 
         public string ApplicationInsightsKey { get; set; }
 
+        public string AdminIdentities { get; set; }
```
Consider deserializing directly as a string array, e.g.
https://github.com/NuGet/NuGetGallery/blob/d7aaa8dd8aa213f77ca61d88eee1c4c7df2ed2d9/src/NuGetGallery.Services/Configuration/AppConfiguration.cs#L39

drewgillies replied:
I did consider this, imported the `NuGet.Services.Configuration` package to get the attribute, and it was ignored by the deserialization. Rather than spend time digging into why deserialization ignored it and returned null (for workaround code), I thought it more expedient to get this up and running with my own split statement.

joelverhagen commented on src/NuGet.Status/Controllers/AdminController.cs:
```diff
@@ -3,94 +3,115 @@
         public const string UpdatedTempDataKey = "Updated";
 
+        private string[] adminIdentities;
+        private string[] AdminIdentities => (adminIdentities = adminIdentities ?? 
```
`readonly`

joelverhagen commented on src/NuGet.Status/Controllers/AdminController.cs:
```diff
@@ -3,94 +3,115 @@
         public const string UpdatedTempDataKey = "Updated";
 
+        private string[] adminIdentities;
+        private string[] AdminIdentities => (adminIdentities = adminIdentities ?? 
```
Maybe just do this in a constructor. I think that's a bit more readable than a lazy-like thing.

skofman1 replied:
what would be the error behavior if the AdminEntities config is malformed? Please make sure we fail fast (during service init with a clear error message

---
In reply to: [513149981](https://github.com/NuGet/NuGet.Services.Status/pull/22#discussion_r513149981) [](ancestors = 513149981)

drewgillies replied:
@skofman1 It's a ';'-delimited string, so the only error would be a false negative where an identity isn't found through wrong use of a delimiter or a typo. Misconfiguring would also include entries not conforming to an email pattern. I could put some intelligence in there to determine whether this is the case if you feel it's appropriate.

skofman1 replied:
i think the check you added is sufficient.

joelverhagen commented on src/NuGet.Status/Controllers/AdminController.cs:
```diff
@@ -3,94 +3,119 @@
         public ActionResult Index()
         {
-            return View(nameof(Index), CachedServiceStatus?.ServiceStatus);
+            return AuthorizedAction(() => 
```
I don't like that you have to remember to call `AuthorizedAction` here. What if we add a new action and forget to wrap the logic in `AuthorizedAction`? Wouldn't that allow users not in the admin identity group to perform the action?

Can we have a custom filter applied to the controller that verifies this instead of putting the logic in the action (method) body?

drewgillies replied:
Interesting idea. I can explore that.

drewgillies replied:
Remember that this is intended to be temporary though. But those are famous last words. :)

drewgillies replied:
Okay, yes, this idea is much nicer. Pushing the change.
